osquery/tables/yara/yara.cpp

```diff
@@ -39,71 +40,221 @@ FLAG(bool,
+  auto yaraParser = getYaraParser();
+  if (isNull(yaraParser)) {
+    return Status::failure("YARA config parser plugin has no pointer");
+  }
+
+  auto& rules = yaraParser->rules();
+  auto sigfiles = queryContext.constraints["sigfile"].getAll(EQUALS);
+
+  for (const auto& file : sigfiles) {
+    // Check if the signature file has been used/ compiled
+    if (rules.count(hashStr(file, YC_FILE)) > 0) {
+      scanContext.insert(std::make_pair(YC_FILE, file));
+      continue;
+    }
+
+    // If this is a relative path append the default yara search path.
+    auto path = (file[0] != '/') ? (kYARAHome + file) : file;
+
+    YR_RULES* tmp_rules = nullptr;
+    auto status = compileSingleFile(path, &tmp_rules);
+    if (!status.ok()) {
+      LOG(WARNING) << "YARA compile error : " << status.toString();
+      continue;
+    }
+
+    // Cache the compiled rules by setting the unique signature file path
+    // as the lookup name. Additional signature file uses will skip the
+    // compile step and be added to the scan context
+    rules[hashStr(file, YC_FILE)] = tmp_rules;
+    scanContext.insert(std::make_pair(YC_FILE, file));
+  }
+
+  return Status::success();
+}
+
+Status genYaraRuleFromString(QueryContext& queryContext,
+                             YaraScanContext& scanContext) {
+  auto yaraParser = getYaraParser();
+  if (isNull(yaraParser)) {
+    return Status::failure("YARA config parser plugin has no pointer");
+  }
+
+  auto& rules = yaraParser->rules();
+  auto sigrules = queryContext.constraints["sigrule"].getAll(EQUALS);
+
+  // Compile signature string and add them to the scan context
+  for (const auto& rule_string : sigrules) {
+    // Check if the signature string has been used/compiled
+    if (rules.count(hashStr(rule_string, YC_RULE)) > 0) {
+      scanContext.insert(std::make_pair(YC_RULE, rule_string));
+      continue;
+    }
+
+    YR_RULES* tmp_rules = nullptr;
+    auto status = compileFromString(rule_string, &tmp_rules);
+    if (!status.ok()) {
+      LOG(WARNING) << "YARA compile error : " << status.toString();
+      continue;
+    }
+
+    // Cache the compiled rules by setting the unique hashed signature
+    // string as the lookup name. Additional signature file uses will
+    // skip the compile step and be added to the scan context
+    rules[hashStr(rule_string, YC_RULE)] = tmp_rules;
+    scanContext.insert(std::make_pair(YC_RULE, rule_string));
   }
+
+  return Status::success();
 }
 
 QueryData genYara(QueryContext& context) {
   QueryData results;
+  YaraScanContext scanContext;
 
-  // Must specify a path constraint and at least one of sig_group or sigfile.
-  auto groups = context.constraints["sig_group"].getAll(EQUALS);
-  auto sigfiles = context.constraints["sigfile"].getAll(EQUALS);
-  if (groups.size() == 0 && sigfiles.size() == 0) {
+  // Initialize yara library
+  auto status = yaraInitilize();
+  if (!status.ok()) {
+    LOG(WARNING) << "Error : " << status.toString();
```

Comment:
Is this a warning or an error?

I recommend `LOG(WARNING) << status.toString();`